src/query/api/v1/handler/fetch_options.go

```diff
@@ -111,19 +113,37 @@ func (b fetchOptionsBuilder) NewFetchOptions(
 				"could not parse metrics type: input=%s, err=%v", str, err)
 			return nil, xhttp.NewParseError(err, http.StatusBadRequest)
 		}
+
 		fetchOpts.RestrictFetchOptions = newOrExistingRestrictFetchOptions(fetchOpts)
-		fetchOpts.RestrictFetchOptions.MetricsType = mt
+		fetchOpts.RestrictFetchOptions.RestrictByType =
+			newOrExistingRestrictFetchOptionsRestrictByType(fetchOpts)
+		fetchOpts.RestrictFetchOptions.RestrictByType.MetricsType = mt
 	}
+
 	if str := req.Header.Get(MetricsStoragePolicyHeader); str != "" {
 		sp, err := policy.ParseStoragePolicy(str)
 		if err != nil {
 			err = fmt.Errorf(
 				"could not parse storage policy: input=%s, err=%v", str, err)
 			return nil, xhttp.NewParseError(err, http.StatusBadRequest)
 		}
+
+		fetchOpts.RestrictFetchOptions = newOrExistingRestrictFetchOptions(fetchOpts)
+		fetchOpts.RestrictFetchOptions.RestrictByType =
+			newOrExistingRestrictFetchOptionsRestrictByType(fetchOpts)
+		fetchOpts.RestrictFetchOptions.RestrictByType.StoragePolicy = sp
+	}
+
+	if str := req.Header.Get(FetchRestrictLabels); str != "" {
+		var opts *storage.RestrictFetchOptions
```

Comment:
nit: Should we call this `storage.QueryOptions` to match the header `QueryOptionsJSONHeader = "M3-Query-Options-JSON"` we landed on?

Reply:
Moving to `M3-Restrict-By-Tags-JSON`